libs/completer/completer.go

```diff
@@ -0,0 +1,66 @@
+package completer
+
+import (
+	"context"
+	"path"
+
+	"github.com/databricks/cli/libs/filer"
+	"github.com/spf13/cobra"
+)
+
+type completer struct {
+	ctx context.Context
+
+	// The filer to use for completing remote or local paths.
+	filer filer.Filer
+
+	// CompletePath will only return directories when onlyDirs is true.
+	onlyDirs bool
+}
+
+// General completer that takes a filer to complete remote paths when TAB-ing through a path.
+func New(ctx context.Context, filer filer.Filer, onlyDirs bool) *completer {
+	return &completer{ctx: ctx, filer: filer, onlyDirs: onlyDirs}
+}
+
+func (c *completer) CompletePath(p string) ([]string, string, cobra.ShellCompDirective, error) {
+	// If the user is TAB-ing their way through a path, the path in `toComplete`
+	// is valid and we should list nested directories.
+	// If the path in `toComplete` is incomplete, however,
+	// then we should list adjacent directories.
+	dirPath := p
+	completions, err := fetchCompletions(c, dirPath)
```

Comment:
Behavioral note: when doing local completion, I find that the completion of a valid directory name without trailing slash completes to adding the trailing slash. This also makes sense here: if you have two directories "dir1" and "dir11", then you don't want to assume "dir1" is a valid directory because the user might want to complete to "dir11". It's an edge case but would even lead to simpler code as we can check for a trailing slash to determine what to do (i.e. use `path.Dir` or not).

Reply:
Good idea, added! `sync` will also have trailing slashes once we move it to use this completer.